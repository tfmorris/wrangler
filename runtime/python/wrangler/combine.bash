#!/bin/bash
rm dw.py
for i in 'wrangler' 'table' 'transform' 'map' 'row' 'filter' 'fill' 'fold' 'drop' 'schema' 'sort' 'split' 'unfold'
do
   cat "$i.py" | grep -v 'from [ a-zA-Z_]* import' >> dw.py
   echo "" >> dw.py
done