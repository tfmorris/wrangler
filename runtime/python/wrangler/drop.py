from transform import Transform
import re
class Drop(Transform):
	def __init__(self):
		super(Drop,self).__init__()
		self['drop'] = True
		self.name = 'drop'
	def apply(self, tables):

		columns = self.get_columns(tables)
		table = tables[0]


		
		if(self['drop']):
			for col in columns:
				table.remove_column(col);


		

