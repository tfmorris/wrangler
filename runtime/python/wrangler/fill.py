from transform import Transform
import re
class Fill(Transform):
	def __init__(self):
		super(Fill,self).__init__()
		self['direction'] = 'down'
		self.name = 'fill'
	def apply(self, tables):

		columns = self.get_columns(tables)
		table = tables[0]

		start_row = 0
		end_row = table.rows()
		direction = self['direction']
		
		row = self['row']
		
		if(direction=='down' or direction=='up'):
			row_range = range(start_row, end_row)
			if(direction=='up'):
				row_range.reverse()
			for col in columns:
				fillValue = None
				for r in row_range:
					v = col[r]
					if(v == None or v==''):
						col[r] = fillValue
					else:
						if(row and row.test(table, r)):
							fillValue = v

		if(direction=='left' or direction=='right'):
			col_range = range(0, len(columns))
			if(direction=='left'):
				col_range.reverse()
			for r in range(start_row, end_row):
				if(row and row.test(table, r)):
					fillValue = None
					for c in col_range:
						v = columns[c][r]
						if(v == None or v==''):
							columns[c][r] = fillValue
						else:
							fillValue = v



