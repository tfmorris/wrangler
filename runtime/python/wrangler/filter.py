from transform import Transform
from row import Row
import re

class Filter(Transform):
	def __init__(self):
		super(Filter,self).__init__()
		self['row'] = Row()
		self.name = 'filter'

	def apply(self, tables):

		columns = self.get_columns(tables)
		table = tables[0]

		filtered_table = table.slice(0, 0)
		row = self['row']
		start_row = 0
		end_row = table.rows()


		for r in range(start_row, end_row):
			if(row.test(table, r)):
				pass
			else:
				for col in table:
					filtered_col = filtered_table[col.name]
					filtered_col.append(col[r])
		
		
		table.clear()
		
		table.append(filtered_table)


