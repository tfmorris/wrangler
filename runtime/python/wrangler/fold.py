from transform import Transform
import re
from table import Column
class Fold(Transform):
	def __init__(self):
		super(Fold,self).__init__()
		self['drop'] = False
		self['keys'] = [-1]
		self.name = 'fold'

	def apply(self, tables):

		columns = self.get_columns(tables)
		table = tables[0]
		
		new_index = 0
		names = [c.name for c in columns]
		rows = table.rows()
		start_row = 0
		end_row = table.rows()
		keys = self['keys']
		# These are the keys to use for the fold...We use the header if the key = -1 otherwise we use the value in the cell

		
		
		key_values = []
		for col in columns:
			col_key_vals = []
			for key in keys:
				val = (col.name if key==-1 else col[key])
				col_key_vals.append(val)
				key_values.append(col_key_vals)
					


		
		foundLeft = False;
		other_cols = []
		

		
		for col in table:

			if(col.name in names):
				if(not foundLeft):
					update_col = col
				foundLeft = True				
			else:
				other_cols.append(col)

		key_cols = [table.add_column('fold', {}) for k in keys]
		value_col = table.add_column('value', {})
		other_cols = [Column(c.name, 't','r') for c in other_cols]

		

		for row in range(start_row, end_row):
			if(not row in keys):
				for k in range(0, len(columns)):
					for c in range(0, len(other_cols)):
						col = other_cols[c]
						
						col[new_index] = table[col.name][row]
					
					for j in range(0, len(key_cols)):
						key_cols[j][new_index] = key_values[k][j]
					
					value_col[new_index] = columns[k][row]
					new_index+=1;
				
			
		



		updateIndex = table.index(update_col) if update_col else 0;



		table.clear()


		key_cols.append(value_col)
		

		for col in other_cols:
			table.insert_column(col, {'index':updateIndex})		
		
		for col in key_cols:
			table.insert_column(col, {})




	
