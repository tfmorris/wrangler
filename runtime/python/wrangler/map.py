import transform

class Map(transform.Transform):
	def __init__(self):
		super(Map,self).__init__()
		
	def apply(self, tables):
		
		columns = self.get_columns(tables)
		table = tables[0]
		
		updater = transform.TableUpdater(self, table, columns)
		

		
		for row in range(0, table.rows()):

			values = list([c[row] for c in columns])
			result = self.transform(values)
			updater.update(row, result)
		
		updater.finish()