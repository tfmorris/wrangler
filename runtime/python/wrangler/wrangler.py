class DataWrangler(object):
	def __init__(self):
		self.transforms = [];
		
	def apply(self, tables):

		for t in self.transforms:
			t.apply(tables);		
		
	def add(self, t):
		self.transforms.append(t)
		return self
		
