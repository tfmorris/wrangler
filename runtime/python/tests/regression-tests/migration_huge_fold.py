from wrangler import dw

w = dw.DataWrangler()

# Split data repeatedly on newline  into  rows
w.add(dw.Split(column=["data"],
              table=0,
              status="active",
              drop=True,
              result="row",
              update=False,
              insert_position="right",
              row=None,
              on="\n",
              before=None,
              after=None,
              ignore_between=None,
              which=1,
              max=0,
              positions=None,
              quote_character=None))

# Split data repeatedly on ','
w.add(dw.Split(column=["data"],
              table=0,
              status="active",
              drop=True,
              result="column",
              update=False,
              insert_position="right",
              row=None,
              on=",",
              before=None,
              after=None,
              ignore_between=None,
              which=1,
              max=0,
              positions=None,
              quote_character=None))

# Delete empty rows
w.add(dw.Filter(column=[],
               table=0,
               status="active",
               drop=False,
               row=dw.Row(column=[],
            table=0,
            status="active",
            drop=False,
            conditions=[dw.Empty(column=[],
              table=0,
              status="active",
              drop=False)])))

# Delete rows 1,2
w.add(dw.Filter(column=[],
               table=0,
               status="active",
               drop=False,
               row=dw.Row(column=[],
            table=0,
            status="active",
            drop=False,
            conditions=[dw.RowIndex(column=[],
                 table=0,
                 status="active",
                 drop=False,
                 indices=[0,1])])))

# Fill  rows 1,2,3 with values from the left
w.add(dw.Fill(column=[],
             table=0,
             status="active",
             drop=False,
             direction="right",
             method="copy",
             row=dw.Row(column=[],
            table=0,
            status="active",
            drop=False,
            conditions=[dw.RowIndex(column=[],
                 table=0,
                 status="active",
                 drop=False,
                 indices=[0,1,2])])))

# Fold split1, split2, split3, split4...  using  1, 2, 3  as keys
w.add(dw.Fold(column=["split1","split2","split3","split4","split5","split6","split7","split8","split9","split10","split11","split12","split13","split14","split15","split16"],
             table=0,
             status="active",
             drop=False,
             keys=[0,1,2]))

w.apply_to_file('migration.csv').print_csv('out.csv')
