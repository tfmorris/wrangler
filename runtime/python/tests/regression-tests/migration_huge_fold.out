split,fold,fold1,fold2,value
"Australia","Men","Native","1995","8.4"
"Australia","Men","Native","2000","6.6"
"Australia","Men","Native","2003","6.0"
"Australia","Men","Native","2004","5.6"
"Australia","Men","Foreign..born","1995","10.6"
"Australia","Men","Foreign..born","2000","6.5"
"Australia","Men","Foreign..born","2003","6.5"
"Australia","Men","Foreign..born","2004","5.5"
"Australia","Women","Native","1995","7.7"
"Australia","Women","Native","2000","5.8"
"Australia","Women","Native","2003","6.1"
"Australia","Women","Native","2004","5.7"
"Australia","Women","Foreign..born","1995","9.6"
"Australia","Women","Foreign..born","2000","7.0"
"Australia","Women","Foreign..born","2003","6.5"
"Australia","Women","Foreign..born","2004","5.6"
"Austria","Men","Native","1995","3.6"
"Austria","Men","Native","2000","4.3"
"Austria","Men","Native","2003","4.4"
"Austria","Men","Native","2004","4.3"
"Austria","Men","Foreign..born","1995","6.6"
"Austria","Men","Foreign..born","2000","8.7"
"Austria","Men","Foreign..born","2003","9.7"
"Austria","Men","Foreign..born","2004","11.2"
"Austria","Women","Native","1995","4.6"
"Austria","Women","Native","2000","4.2"
"Austria","Women","Native","2003","4.0"
"Austria","Women","Native","2004","4.3"
"Austria","Women","Foreign..born","1995","7.3"
"Austria","Women","Foreign..born","2000","7.2"
"Austria","Women","Foreign..born","2003","6.6"
"Austria","Women","Foreign..born","2004","10.7"
"Belgium","Men","Native","1995","6.3"
"Belgium","Men","Native","2000","4.2"
"Belgium","Men","Native","2003","6.0"
"Belgium","Men","Native","2004","5.6"
"Belgium","Men","Foreign..born","1995","16.9"
"Belgium","Men","Foreign..born","2000","14.7"
"Belgium","Men","Foreign..born","2003","18.3"
"Belgium","Men","Foreign..born","2004","14.9"
"Belgium","Women","Native","1995","11.2"
"Belgium","Women","Native","2000","7.4"
"Belgium","Women","Native","2003","6.9"
"Belgium","Women","Native","2004","7.5"
"Belgium","Women","Foreign..born","1995","23.8"
"Belgium","Women","Foreign..born","2000","17.5"
"Belgium","Women","Foreign..born","2003","17.3"
"Belgium","Women","Foreign..born","2004","15.0"
"Canada","Men","Native","1995","8.6"
"Canada","Men","Native","2000","5.7"
"Canada","Men","Native","2003","6.5"
"Canada","Men","Native","2004",".."
"Canada","Men","Foreign..born","1995","10.4"
"Canada","Men","Foreign..born","2000","6.1"
"Canada","Men","Foreign..born","2003","7.8"
"Canada","Men","Foreign..born","2004",".."
"Canada","Women","Native","1995","9.8"
"Canada","Women","Native","2000","6.2"
"Canada","Women","Native","2003","5.9"
"Canada","Women","Native","2004",".."
"Canada","Women","Foreign..born","1995","13.3"
"Canada","Women","Foreign..born","2000","8.7"
"Canada","Women","Foreign..born","2003","9.9"
"Canada","Women","Foreign..born","2004",".."
"Czech Republic","Men","Native","1995",".."
"Czech Republic","Men","Native","2000",".."
"Czech Republic","Men","Native","2003","5.8"
"Czech Republic","Men","Native","2004","7.0"
"Czech Republic","Men","Foreign..born","1995",".."
"Czech Republic","Men","Foreign..born","2000",".."
"Czech Republic","Men","Foreign..born","2003","9.0"
"Czech Republic","Men","Foreign..born","2004","12.4"
"Czech Republic","Women","Native","1995",".."
"Czech Republic","Women","Native","2000",".."
"Czech Republic","Women","Native","2003","9.6"
"Czech Republic","Women","Native","2004","9.6"
"Czech Republic","Women","Foreign..born","1995",".."
"Czech Republic","Women","Foreign..born","2000",".."
"Czech Republic","Women","Foreign..born","2003","15.7"
"Czech Republic","Women","Foreign..born","2004","13.5"
"Denmark","Men","Native","1995","6.4"
"Denmark","Men","Native","2000","3.4"
"Denmark","Men","Native","2003","3.8"
"Denmark","Men","Native","2004","4.6"
"Denmark","Men","Foreign..born","1995","20.5"
"Denmark","Men","Foreign..born","2000","9.5"
"Denmark","Men","Foreign..born","2003","8.8"
"Denmark","Men","Foreign..born","2004","11.8"
"Denmark","Women","Native","1995","8.4"
"Denmark","Women","Native","2000","4.3"
"Denmark","Women","Native","2003","4.2"
"Denmark","Women","Native","2004","5.2"
"Denmark","Women","Foreign..born","1995","20.7"
"Denmark","Women","Foreign..born","2000","9.6"
"Denmark","Women","Foreign..born","2003","8.7"
"Denmark","Women","Foreign..born","2004","12.7"
"Finland","Men","Native","1995","17.7"
"Finland","Men","Native","2000","10.3"
"Finland","Men","Native","2003","10.9"
"Finland","Men","Native","2004","9.9"
"Finland","Men","Foreign..born","1995",".."
"Finland","Men","Foreign..born","2000",".."
"Finland","Men","Foreign..born","2003","18.4"
"Finland","Men","Foreign..born","2004","21.3"
"Finland","Women","Native","1995","16.1"
"Finland","Women","Native","2000","12.0"
"Finland","Women","Native","2003","9.7"
"Finland","Women","Native","2004","10.2"
"Finland","Women","Foreign..born","1995",".."
"Finland","Women","Foreign..born","2000",".."
"Finland","Women","Foreign..born","2003","20.0"
"Finland","Women","Foreign..born","2004","25.3"
"France","Men","Native","1995","9.1"
"France","Men","Native","2000","7.7"
"France","Men","Native","2003","7.3"
"France","Men","Native","2004","8.0"
"France","Men","Foreign..born","1995","16.6"
"France","Men","Foreign..born","2000","14.5"
"France","Men","Foreign..born","2003","15.4"
"France","Men","Foreign..born","2004","13.8"
"France","Women","Native","1995","13.6"
"France","Women","Native","2000","11.3"
"France","Women","Native","2003","9.2"
"France","Women","Native","2004","9.9"
"France","Women","Foreign..born","1995","19.0"
"France","Women","Foreign..born","2000","19.7"
"France","Women","Foreign..born","2003","16.4"
"France","Women","Foreign..born","2004","17.4"
"Germany","Men","Native","1995",".."
"Germany","Men","Native","2000","6.9"
"Germany","Men","Native","2003","9.3"
"Germany","Men","Native","2004","10.3"
"Germany","Men","Foreign..born","1995",".."
"Germany","Men","Foreign..born","2000","12.9"
"Germany","Men","Foreign..born","2003","16.9"
"Germany","Men","Foreign..born","2004","18.3"
"Germany","Women","Native","1995",".."
"Germany","Women","Native","2000","8.0"
"Germany","Women","Native","2003","8.8"
"Germany","Women","Native","2004","9.6"
"Germany","Women","Foreign..born","1995",".."
"Germany","Women","Foreign..born","2000","12.1"
"Germany","Women","Foreign..born","2003","14.0"
"Germany","Women","Foreign..born","2004","15.2"
"Greece","Men","Native","1995","6.1"
"Greece","Men","Native","2000","7.4"
"Greece","Men","Native","2003","5.8"
"Greece","Men","Native","2004","6.5"
"Greece","Men","Foreign..born","1995","14.0"
"Greece","Men","Foreign..born","2000","9.5"
"Greece","Men","Foreign..born","2003","6.5"
"Greece","Men","Foreign..born","2004","6.5"
"Greece","Women","Native","1995","13.7"
"Greece","Women","Native","2000","16.6"
"Greece","Women","Native","2003","13.7"
"Greece","Women","Native","2004","15.7"
"Greece","Women","Foreign..born","1995","20.8"
"Greece","Women","Foreign..born","2000","21.1"
"Greece","Women","Foreign..born","2003","15.7"
"Greece","Women","Foreign..born","2004","19.1"
"Hungary","Men","Native","1995",".."
"Hungary","Men","Native","2000","7.3"
"Hungary","Men","Native","2003","6.2"
"Hungary","Men","Native","2004","5.9"
"Hungary","Men","Foreign..born","1995",".."
"Hungary","Men","Foreign..born","2000",".."
"Hungary","Men","Foreign..born","2003",".."
"Hungary","Men","Foreign..born","2004","2.0"
"Hungary","Women","Native","1995",".."
"Hungary","Women","Native","2000","5.8"
"Hungary","Women","Native","2003","5.4"
"Hungary","Women","Native","2004","5.9"
"Hungary","Women","Foreign..born","1995",".."
"Hungary","Women","Foreign..born","2000",".."
"Hungary","Women","Foreign..born","2003",".."
"Hungary","Women","Foreign..born","2004","6.4"
"Ireland","Men","Native","1995","12.0"
"Ireland","Men","Native","2000","4.4"
"Ireland","Men","Native","2003","4.8"
"Ireland","Men","Native","2004","4.9"
"Ireland","Men","Foreign..born","1995","16.8"
"Ireland","Men","Foreign..born","2000",".."
"Ireland","Men","Foreign..born","2003","6.6"
"Ireland","Men","Foreign..born","2004","6.7"
"Ireland","Women","Native","1995","11.9"
"Ireland","Women","Native","2000","4.2"
"Ireland","Women","Native","2003","3.8"
"Ireland","Women","Native","2004","3.7"
"Ireland","Women","Foreign..born","1995","15.4"
"Ireland","Women","Foreign..born","2000",".."
"Ireland","Women","Foreign..born","2003","6.0"
"Ireland","Women","Foreign..born","2004","5.3"
"Italy","Men","Native","1995","9.3"
"Italy","Men","Native","2000","8.4"
"Italy","Men","Native","2003","7.0"
"Italy","Men","Native","2004","6.4"
"Italy","Men","Foreign..born","1995",".."
"Italy","Men","Foreign..born","2000","6.5"
"Italy","Men","Foreign..born","2003","3.8"
"Italy","Men","Foreign..born","2004","6.2"
"Italy","Women","Native","1995","16.3"
"Italy","Women","Native","2000","14.9"
"Italy","Women","Native","2003","12.0"
"Italy","Women","Native","2004","10.1"
"Italy","Women","Foreign..born","1995","23.5"
"Italy","Women","Foreign..born","2000","21.2"
"Italy","Women","Foreign..born","2003","10.5"
"Italy","Women","Foreign..born","2004","13.2"
"Luxembourg","Men","Native","1995",".."
"Luxembourg","Men","Native","2000",".."
"Luxembourg","Men","Native","2003","2.3"
"Luxembourg","Men","Native","2004","2.4"
"Luxembourg","Men","Foreign..born","1995",".."
"Luxembourg","Men","Foreign..born","2000",".."
"Luxembourg","Men","Foreign..born","2003","3.9"
"Luxembourg","Men","Foreign..born","2004","4.4"
"Luxembourg","Women","Native","1995",".."
"Luxembourg","Women","Native","2000",".."
"Luxembourg","Women","Native","2003","3.6"
"Luxembourg","Women","Native","2004","4.5"
"Luxembourg","Women","Foreign..born","1995",".."
"Luxembourg","Women","Foreign..born","2000",".."
"Luxembourg","Women","Foreign..born","2003","5.9"
"Luxembourg","Women","Foreign..born","2004","9.6"
"Netherlands","Men","Native","1995","4.9"
"Netherlands","Men","Native","2000","1.8"
"Netherlands","Men","Native","2003","2.8"
"Netherlands","Men","Native","2004","3.6"
"Netherlands","Men","Foreign..born","1995","19.5"
"Netherlands","Men","Foreign..born","2000","5.4"
"Netherlands","Men","Foreign..born","2003","9.1"
"Netherlands","Men","Foreign..born","2004","10.3"
"Netherlands","Women","Native","1995","7.7"
"Netherlands","Women","Native","2000","3.0"
"Netherlands","Women","Native","2003","3.2"
"Netherlands","Women","Native","2004","4.3"
"Netherlands","Women","Foreign..born","1995","19.8"
"Netherlands","Women","Foreign..born","2000","7.6"
"Netherlands","Women","Foreign..born","2003","8.6"
"Netherlands","Women","Foreign..born","2004","10.6"
"Norway","Men","Native","1995",".."
"Norway","Men","Native","2000","3.4"
"Norway","Men","Native","2003","4.0"
"Norway","Men","Native","2004","4.3"
"Norway","Men","Foreign..born","1995",".."
"Norway","Men","Foreign..born","2000","6.8"
"Norway","Men","Foreign..born","2003","11.1"
"Norway","Men","Foreign..born","2004","8.9"
"Norway","Women","Native","1995",".."
"Norway","Women","Native","2000","3.2"
"Norway","Women","Native","2003","3.8"
"Norway","Women","Native","2004","3.7"
"Norway","Women","Foreign..born","1995",".."
"Norway","Women","Foreign..born","2000",".."
"Norway","Women","Foreign..born","2003",".."
"Norway","Women","Foreign..born","2004","7.3"
"Portugal","Men","Native","1995","6.6"
"Portugal","Men","Native","2000","3.1"
"Portugal","Men","Native","2003","5.3"
"Portugal","Men","Native","2004","5.7"
"Portugal","Men","Foreign..born","1995",".."
"Portugal","Men","Foreign..born","2000","3.9"
"Portugal","Men","Foreign..born","2003","7.9"
"Portugal","Men","Foreign..born","2004","9.8"
"Portugal","Women","Native","1995","7.8"
"Portugal","Women","Native","2000","4.9"
"Portugal","Women","Native","2003","7.4"
"Portugal","Women","Native","2004","7.4"
"Portugal","Women","Foreign..born","1995",".."
"Portugal","Women","Foreign..born","2000","5.4"
"Portugal","Women","Foreign..born","2003","10.4"
"Portugal","Women","Foreign..born","2004","9.6"
"Slovak Republic","Men","Native","1995",".."
"Slovak Republic","Men","Native","2000",".."
"Slovak Republic","Men","Native","2003","17.0"
"Slovak Republic","Men","Native","2004","17.8"
"Slovak Republic","Men","Foreign..born","1995",".."
"Slovak Republic","Men","Foreign..born","2000",".."
"Slovak Republic","Men","Foreign..born","2003",".."
"Slovak Republic","Men","Foreign..born","2004","17.9"
"Slovak Republic","Women","Native","1995",".."
"Slovak Republic","Women","Native","2000",".."
"Slovak Republic","Women","Native","2003","17.2"
"Slovak Republic","Women","Native","2004","19.5"
"Slovak Republic","Women","Foreign..born","1995",".."
"Slovak Republic","Women","Foreign..born","2000",".."
"Slovak Republic","Women","Foreign..born","2003",".."
"Slovak Republic","Women","Foreign..born","2004","30.5"
"Spain","Men","Native","1995","18.0"
"Spain","Men","Native","2000","9.5"
"Spain","Men","Native","2003","7.9"
"Spain","Men","Native","2004","7.8"
"Spain","Men","Foreign..born","1995","24.4"
"Spain","Men","Foreign..born","2000","12.4"
"Spain","Men","Foreign..born","2003","10.4"
"Spain","Men","Foreign..born","2004","11.4"
"Spain","Women","Native","1995","30.5"
"Spain","Women","Native","2000","20.5"
"Spain","Women","Native","2003","15.8"
"Spain","Women","Native","2004","15.1"
"Spain","Women","Foreign..born","1995","30.5"
"Spain","Women","Foreign..born","2000","20.7"
"Spain","Women","Foreign..born","2003","17.2"
"Spain","Women","Foreign..born","2004","17.1"
"Sweden","Men","Native","1995","7.9"
"Sweden","Men","Native","2000","5.1"
"Sweden","Men","Native","2003","5.2"
"Sweden","Men","Native","2004","6.2"
"Sweden","Men","Foreign..born","1995","24.8"
"Sweden","Men","Foreign..born","2000","12.3"
"Sweden","Men","Foreign..born","2003","12.7"
"Sweden","Men","Foreign..born","2004","14.2"
"Sweden","Women","Native","1995","6.6"
"Sweden","Women","Native","2000","4.2"
"Sweden","Women","Native","2003","4.4"
"Sweden","Women","Native","2004","5.2"
"Sweden","Women","Foreign..born","1995","18.5"
"Sweden","Women","Foreign..born","2000","10.8"
"Sweden","Women","Foreign..born","2003","9.5"
"Sweden","Women","Foreign..born","2004","12.6"
"Switzerland","Men","Native","1995",".."
"Switzerland","Men","Native","2000",".."
"Switzerland","Men","Native","2003","2.8"
"Switzerland","Men","Native","2004","2.9"
"Switzerland","Men","Foreign..born","1995",".."
"Switzerland","Men","Foreign..born","2000",".."
"Switzerland","Men","Foreign..born","2003","7.2"
"Switzerland","Men","Foreign..born","2004","7.5"
"Switzerland","Women","Native","1995",".."
"Switzerland","Women","Native","2000",".."
"Switzerland","Women","Native","2003","3.0"
"Switzerland","Women","Native","2004","3.4"
"Switzerland","Women","Foreign..born","1995",".."
"Switzerland","Women","Foreign..born","2000",".."
"Switzerland","Women","Foreign..born","2003","9.1"
"Switzerland","Women","Foreign..born","2004","9.2"
"United Kingdom","Men","Native","1995","9.9"
"United Kingdom","Men","Native","2000","5.9"
"United Kingdom","Men","Native","2003","5.2"
"United Kingdom","Men","Native","2004","4.7"
"United Kingdom","Men","Foreign..born","1995","14.2"
"United Kingdom","Men","Foreign..born","2000","9.6"
"United Kingdom","Men","Foreign..born","2003","8.1"
"United Kingdom","Men","Foreign..born","2004","7.3"
"United Kingdom","Women","Native","1995","6.7"
"United Kingdom","Women","Native","2000","4.6"
"United Kingdom","Women","Native","2003","3.9"
"United Kingdom","Women","Native","2004","3.9"
"United Kingdom","Women","Foreign..born","1995","10.9"
"United Kingdom","Women","Foreign..born","2000","7.8"
"United Kingdom","Women","Foreign..born","2003","6.3"
"United Kingdom","Women","Foreign..born","2004","7.3"
"United States","Men","Native","1995","6.2"
"United States","Men","Native","2000","4.5"
"United States","Men","Native","2003","7.0"
"United States","Men","Native","2004","6.9"
"United States","Men","Foreign..born","1995","7.9"
"United States","Men","Foreign..born","2000","4.5"
"United States","Men","Foreign..born","2003","7.2"
"United States","Men","Foreign..born","2004","5.8"
"United States","Women","Native","1995","5.3"
"United States","Women","Native","2000","4.2"
"United States","Women","Native","2003","5.7"
"United States","Women","Native","2004","5.5"
"United States","Women","Foreign..born","1995","8.2"
"United States","Women","Foreign..born","2000","5.5"
"United States","Women","Foreign..born","2003","8.0"
"United States","Women","Foreign..born","2004","6.8"
